include/cuco/detail/static_reduction_map.inl

```diff
@@ -0,0 +1,500 @@
+}
+
+template <typename ReductionOp,
+          typename Key,
+          typename Value,
+          cuda::thread_scope Scope,
+          typename Allocator>
+static_reduction_map<ReductionOp, Key, Value, Scope, Allocator>::~static_reduction_map()
+{
+  std::allocator_traits<slot_allocator_type>::deallocate(slot_allocator_, slots_, capacity_);
+}
+
+template <typename ReductionOp,
+          typename Key,
+          typename Value,
+          cuda::thread_scope Scope,
+          typename Allocator>
+template <typename InputIt, typename Hash, typename KeyEqual>
+void static_reduction_map<ReductionOp, Key, Value, Scope, Allocator>::insert(
+  InputIt first, InputIt last, cudaStream_t stream, Hash hash, KeyEqual key_equal)
+{
+  auto const num_keys = std::distance(first, last);
+  if (num_keys == 0) { return; }
+
+  auto constexpr block_size = 128;
+  auto constexpr stride     = 1;
+  auto constexpr tile_size  = 4;
+  auto const grid_size = (tile_size * num_keys + stride * block_size - 1) / (stride * block_size);
+  auto view            = get_device_mutable_view();
+
+  detail::insert<block_size, tile_size>
+    <<<grid_size, block_size, 0, stream>>>(first, first + num_keys, view, hash, key_equal);
+}
+
+template <typename ReductionOp,
+          typename Key,
+          typename Value,
+          cuda::thread_scope Scope,
+          typename Allocator>
+template <typename InputIt, typename OutputIt, typename Hash, typename KeyEqual>
+void static_reduction_map<ReductionOp, Key, Value, Scope, Allocator>::find(InputIt first,
+                                                                           InputIt last,
+                                                                           OutputIt output_begin,
+                                                                           cudaStream_t stream,
+                                                                           Hash hash,
+                                                                           KeyEqual key_equal)
+{
+  auto num_keys = std::distance(first, last);
+  if (num_keys == 0) { return; }
+
+  auto constexpr block_size = 128;
+  auto constexpr stride     = 1;
+  auto constexpr tile_size  = 4;
+  auto const grid_size = (tile_size * num_keys + stride * block_size - 1) / (stride * block_size);
+  auto const view      = get_device_view();
+
+  detail::find<block_size, tile_size>
+    <<<grid_size, block_size, 0, stream>>>(first, last, output_begin, view, hash, key_equal);
+}
+
+namespace detail {
+template <typename Key, typename Value>
+struct slot_to_tuple {
+  template <typename S>
+  __device__ thrust::tuple<Key, Value> operator()(S const& s)
+  {
+    return thrust::tuple<Key, Value>(s.first, s.second);
+  }
+};
+
+template <typename Key>
+struct slot_is_filled {
+  Key empty_key_sentinel;
+  template <typename S>
+  __device__ bool operator()(S const& s)
+  {
+    return thrust::get<0>(s) != empty_key_sentinel;
+  }
+};
+}  // namespace detail
+
```

Comment:
```suggestion
```
@sleeepyjack This can be removed since I've moved them to `detail/utils.cuh` in https://github.com/NVIDIA/cuCollections/pull/150